src/mirrors.rs

```diff
@@ -0,0 +1,240 @@
+use log::{error, info, trace, warn};
+
+const MAX_CONNECT_RETRIES: u32 = 5;
+const MAX_SEND_RETRIES: u32 = 3;
+
+pub struct MirroredClient {
+    address: Address,
+    user: User,
+    database: String,
+    bytes_rx: Receiver<Bytes>,
+    disconnect_rx: Receiver<()>,
+    successful_sends_without_recv: u32,
+}
+
+impl MirroredClient {
+    async fn recv_if_neccessary(&mut self, server: &mut Server) -> Result<(), Error> {
+        /* We only receive a response from the server if we have successfully sent
+          5 messages on the current connection. We also send a flush message to gaurantee
+          a server response.
+        */
+        if self.successful_sends_without_recv >= 5 {
+            // We send a flush message to gaurantee a server response
+            server.send(&flush()).await?;
+            server.recv().await?;
+            self.successful_sends_without_recv = 0;
+            trace!(
+                "Received a response from mirror server {:?}",
+                server.address()
+            );
+        }
+        Ok(())
+    }
+
+    async fn connect_with_retries(&mut self, server_id: i32) -> Option<Server> {
+        let mut delay = Duration::from_secs(0);
+        let min_backoff = Duration::from_millis(100);
+        let max_backoff = Duration::from_secs(5);
+        let mut retries = 0;
+
+        loop {
+            if let Some(tmp_server) = self.connect(server_id).await {
+                if !tmp_server.is_bad() {
+                    break Some(tmp_server);
+                }
+            }
+            delay = max(min_backoff, delay);
+            delay = min(max_backoff, delay * 2);
+            retries += 1;
+            if retries > MAX_CONNECT_RETRIES {
+                break None;
+            }
+            sleep(delay).await;
+        }
+    }
+
+    async fn connect(&mut self, server_id: i32) -> Option<Server> {
+        self.successful_sends_without_recv = 0;
+        let stats = get_reporter();
+        stats.server_register(
+            server_id,
+            self.address.id,
+            self.address.name(),
+            self.database.clone(),
+            self.address.username.clone(),
+        );
+        stats.server_login(server_id);
+
+        match Server::startup(
+            server_id,
+            &self.address.clone(),
+            &self.user.clone(),
+            self.database.as_str(),
+            ClientServerMap::default(),
+            get_reporter(),
+        )
+        .await
+        {
+            Ok(conn) => {
+                stats.server_idle(server_id);
+                Some(conn)
+            }
+            Err(_) => {
+                stats.server_disconnecting(server_id);
+                None
+            }
+        }
+    }
+
+    pub fn start(mut self, server_id: i32) {
+        tokio::spawn(async move {
+            let address = self.address.clone();
+            let mut server_optional: Option<Server> = None;
+            loop {
+                tokio::select! {
```

Comment:
nit: It may be possible to refactor this outer loop into it's own function returning an Option so we can use the `?` operator in that function. The goal being to remove the calls to `unwrap`. I realize we're breaking / continuing using `if option.is_none()`, but if someone refactors later and removes the break / continue this could start failing in production.

Reply:
I simplified the logic a ton, I think your concern should be addressed now.

Reply:
This looks great. 🙇